### PR TITLE
Evaluate variables stage updates

### PR DESCRIPTION
**Pull request by christopherthielen**
Everyone, meet the new Evaluate Variables stage:

Nov-08-2019 16-12-32

Lots of changes here, this is a full rewrite:
- Swap out MapEditor in favor of FormikFormFields.  Note how the `label` prop is a nested `FormikFormField` 🤷‍♂ 
- Added Variable Preview and Validation via `SpelInput`
- Better field validation via Formik
- Added a Preview Execution/Stage picker.  This enables us to evaluate the previews against a previous execution and at the right spot in that execution.
- Prepopulate new stages with a blank row

Future work:

- Rewire autocomplete feature

- There are some backend changes coming from @marchello2000 (SpEL v4) that will cause these variables to get evaluated in order, enabling us to provide clear variable dependency rules.  For example, `host` variable is `myapp.${parameters.region}.${parameters.account}.netflix.net` and `delete_endpoint` is `https://${host}/delete`.  We will need to update this stage and SpelInput to preview all the variables in one shot instead of separate API calls.

Future thoughts around SpEL:
- I'd like push users to put all spel into Evaluate Variables stage
- We should only surface the fancy new spel preview UI in the Evaluate Variables stage, as a 🥕 
- We should have a UI for all stage config fields that allow you to pick a variable from a dropdown list
- We should have a clippy mascot that says "It looks like you have inline SpEL in your stage config! Would you like to move them to an Evaluate Variables stage?"
- FormikFormField and FormField components should be responsible for autodetecting SpEL expressions in values and rendering a SpelInput.  This would enable spel support globally.
- FormikFormField and FormField components should render a simple toggle that allows users to switch from (for example) a SelectInput to a VariablePicker input (which doesn't exist yet)

There are some mockups regarding these future thoughts done by @gcomstock available here: https://gallery.io/projects/MCHbtQVoQ2HCZcM1NWkzJjWA/files/MCEvKPS5RTlTG6LlhmLflRIS

### CODE REVIEW

File: app/scripts/modules/core/src/pipeline/config/stages/evaluateVariables/EvaluateVariablesStageConfig.tsx [hunk 1]
```diff
@@ -1,48 +1,232 @@
...
-      <div className="form-horizontal">
-        <StageConfigField label="Variables to evaluate">
-          <MapEditor
-            model={variablesObject}
-            allowEmpty={true}
-            onChange={(v: any) => this.mapChanged('variables', v)}
-            pipeline={pipeline}
-            valueCanContainSpel={true}
+const variableNameValidator: IValidator = (val: string, label: string) =>
+  !val.match(/^[a-zA-Z0-9_]+$/) && errorMessage(`${label} should consist only of letters, numbers, or underscore`);
```
Comment (marchello2000) [inline app/scripts/modules/core/src/pipeline/config/stages/evaluateVariables/EvaluateVariablesStageConfig.tsx]: there are 1 or 2 pipelines that `-` but prob makes sense to not allow it. Also, you should make that variable name doesn't start with a number - that will cause subsequent usages of that variable to fail
```suggestion
  !val.match(/^[a-zA-Z_][a-zA-Z0-9_]+$/) && errorMessage(`${label} should consist only of letters, numbers, or underscore`);
```

File: app/scripts/modules/core/src/pipeline/config/stages/evaluateVariables/EvaluateVariablesStageConfig.tsx [hunk 2]
```diff
@@ -1,48 +1,232 @@
...
+        }
+      }}
+      touched={touchedOverride}
+      label={variableNameInputAsLabel}
+      actions={actions}
+      fastField={false}
+      input={inputProps => (
+        <SpelInput
+          {...inputProps}
+          placeholder="Variable value, e.g. ${trigger.properties.myproperty}"
```
Comment (marchello2000) [inline app/scripts/modules/core/src/pipeline/config/stages/evaluateVariables/EvaluateVariablesStageConfig.tsx]: nit: i would prefer a better/more real example, eg.
```suggestion
          placeholder="Variable value, e.g. ${trigger.buildInfo.number}"
```

File: app/scripts/modules/core/src/pipeline/config/stages/evaluateVariables/ExecutionAndStagePicker.tsx [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+    setStageId(exactStage || findCloseStageFromExecution(pipeline, execution, pipelineStage));
+  }, [pipeline, pipelineStage, execution]);
+
+  // When the chosen execution or stage changes, notify the parent
+  React.useEffect(() => {
+    onChange({ executionLabel: executionLabel(execution), executionId: execution && execution.id, stageId });
+  }, [execution, stageId]);
+
+  const stageOptions: Option[] = stages ? stages.map(x => ({ label: x.name, value: x.id })) : [];
+  const executionOptions: Array<Option<IExecution>> = executions ? executions.map(value => ({ value })) : [];
```
Comment (marchello2000) [inline app/scripts/modules/core/src/pipeline/config/stages/evaluateVariables/ExecutionAndStagePicker.tsx]: what if there are no prior executions? is there some helptext/indication that preview won't work

File: app/scripts/modules/core/src/pipeline/config/stages/evaluateVariables/EvaluateVariablesStageConfig.tsx [hunk 4]
```diff
@@ -1,48 +1,155 @@
...
+function EvaluateVariablesStageForm(props: IFormikStageConfigInjectedProps) {
+  const { formik } = props;
+
+  React.useEffect(() => {
+    const { variables = [] } = props.formik.values;
+    const initialTouched = variables.map(() => ({ key: true, value: true }));
+    formik.setTouched({ variables: initialTouched } as any);
+  }, []);
+
+  const [deleteCount, setDeleteCount] = React.useState(0);
```
Comment (maggieneterval) [inline app/scripts/modules/core/src/pipeline/config/stages/evaluateVariables/EvaluateVariablesStageConfig.tsx]: Would it be possible to generate and persist some UUID for each variable so we don't need this? :pray:

File: app/scripts/modules/core/src/pipeline/config/stages/evaluateVariables/ExecutionAndStagePicker.tsx (hunk 3 above)
Comment (maggieneterval) [reply app/scripts/modules/core/src/pipeline/config/stages/evaluateVariables/ExecutionAndStagePicker.tsx]: +1 to some help text when there are no prior executions if possible!

File: app/scripts/modules/core/src/pipeline/config/stages/evaluateVariables/EvaluateVariablesStageConfig.tsx [hunk 5]
```diff
@@ -1,48 +1,232 @@
...
-    return (
-      <div className="form-horizontal">
-        <StageConfigField label="Variables to evaluate">
-          <MapEditor
-            model={variablesObject}
-            allowEmpty={true}
-            onChange={(v: any) => this.mapChanged('variables', v)}
-            pipeline={pipeline}
-            valueCanContainSpel={true}
+const variableNameValidator: IValidator = (val: string, label: string) =>
```
Comment (caseyhebebrand) [inline app/scripts/modules/core/src/pipeline/config/stages/evaluateVariables/EvaluateVariablesStageConfig.tsx]: If there are other use cases for alphanumeric, snake case string validation then we should move this to the general validator file.

Comment (erikmunson) [reply app/scripts/modules/core/src/pipeline/config/stages/evaluateVariables/EvaluateVariablesStageConfig.tsx]: +1 honestly even if there aren't use cases elsewhere we can think of off the top of our heads right now, I'd like to see this in the core set of validators as it's a pretty generic requirement (might even be applicable to things like app/stack/detail on infra?)

File: app/scripts/modules/core/src/pipeline/config/stages/evaluateVariables/EvaluateVariablesStageConfig.tsx [hunk 6]
```diff
@@ -1,48 +1,232 @@
...
+        .withValidators(variableNameValidator, duplicateKeyValidator);
+      item.field('value', 'Expression').required();
+    }),
+  );
+  return formValidator.validateForm();
+}
+
+function PreviewConfiguration(props: IExecutionAndStagePickerProps) {
+  return (
+    <div style={{ marginLeft: '16px', marginRight: '16px' }}>
```
Comment (erikmunson) [inline app/scripts/modules/core/src/pipeline/config/stages/evaluateVariables/EvaluateVariablesStageConfig.tsx]: I'd like to see stuff like this become `sp-margin-l-xaxis` so we can keep spacing a bit less loosey goosey. You happened to pick the value we use for those classes in this case, but still :smile:

File: app/scripts/modules/core/src/pipeline/config/stages/evaluateVariables/EvaluateVariablesStageConfig.tsx (hunk 4 above)
Comment (erikmunson) [reply app/scripts/modules/core/src/pipeline/config/stages/evaluateVariables/EvaluateVariablesStageConfig.tsx]: theoretically we don't even need to persist it, as long as it's stable per-mount.

**Pull request merged**